streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatterTest.java

```diff
@@ -0,0 +1,316 @@
+                    // found timestamp conflict. use original timestamp
+                    updatedRecordTimestamp = testCase.records.get(updateIdx).timestamp;
+                }
+            }
+            final TestRecord updatedRecord = new TestRecord("updated".getBytes(), updatedRecordTimestamp);
+
+            final SegmentValue segmentValue = buildSegmentWithInsertLatest(testCase);
+
+            // updateRecord() first requires a call to find()
+            segmentValue.find(testCase.records.get(updateIdx).timestamp, false);
+            segmentValue.updateRecord(updatedRecord.timestamp, updatedRecord.value, updateIdx);
+
+            // create expected results
+            final List<TestRecord> expectedRecords = new ArrayList<>(testCase.records);
+            expectedRecords.remove(updateIdx);
+            expectedRecords.add(updateIdx, updatedRecord);
+
+            verifySegmentContents(segmentValue, new TestCase("expected", testCase.nextTimestamp, expectedRecords));
+        }
+    }
+
+    @Test
+    public void shouldFindByTimestamp() {
+        if (testCase.isDegenerate) {
+            // cannot find() on degenerate segment
+            return;
+        }
+
+        final SegmentValue segmentValue = buildSegmentWithInsertLatest(testCase);
+
+        // build expected mapping from timestamp -> record
+        final Map<Long, Integer> expectedRecordIndices = new HashMap<>();
+        for (int recordIdx = testCase.records.size() - 1; recordIdx >= 0; recordIdx--) {
+            if (recordIdx < testCase.records.size() - 1) {
+                expectedRecordIndices.put(testCase.records.get(recordIdx).timestamp - 1, recordIdx + 1);
+            }
+            if (recordIdx > 0) {
+                expectedRecordIndices.put(testCase.records.get(recordIdx).timestamp + 1, recordIdx);
+            }
+            expectedRecordIndices.put(testCase.records.get(recordIdx).timestamp, recordIdx);
+        }
+
+        // verify results
+        for (final Map.Entry<Long, Integer> entry : expectedRecordIndices.entrySet()) {
+            final TestRecord expectedRecord = testCase.records.get(entry.getValue());
+            final long expectedValidTo = entry.getValue() == 0 ? testCase.nextTimestamp : testCase.records.get(entry.getValue() - 1).timestamp;
+
+            final SegmentSearchResult result = segmentValue.find(entry.getKey(), true);
+
+            assertThat(result.index(), equalTo(entry.getValue()));
+            assertThat(result.value(), equalTo(expectedRecord.value));
+            assertThat(result.validFrom(), equalTo(expectedRecord.timestamp));
+            assertThat(result.validTo(), equalTo(expectedValidTo));
+        }
+
+        // verify exception when timestamp is out of range
+        assertThrows(IllegalArgumentException.class, () -> segmentValue.find(testCase.nextTimestamp, false));
+        assertThrows(IllegalArgumentException.class, () -> segmentValue.find(testCase.nextTimestamp + 1, false));
```

Comment:
Why do we need to test this one? Seems the line above using `nextTimestamp` is sufficient?

Reply:
You're right, they're redundant. I added both to make it clear that it's invalid to find not just at the `nextTimestamp` itself, but for all timestamps greater as well. I can remove this if its inclusion is more confusing than it's worth.